enhancements/ingress/custom-route-configuration.md

```diff
@@ -0,0 +1,242 @@
+serving certificates for.
+
+## Motivation
+
+Some customers do not allow wildcard serving certificates for the openshift ingress to use, so they have to provide individual
+serving certificates to each component.
+Those customers need a way to see all the routes they need to customize and a way to configure them.
+A stock installation has the following routes which may need customization of both names and serving cert/keys
+1. openshift-authentication   oauth-openshift
+2. openshift-console          console
+3. openshift-console          downloads
+4. openshift-monitoring       alertmanager-main
+5. openshift-monitoring       grafana
+6. openshift-monitoring       prometheus-k8s
+7. openshift-monitoring       thanos-querier
+
+It is just as easy to create a generic solution that provides one stop shopping for cluster-admins as it is to produce
+a series of one-off solutions.
+This proposes a single point of configuration because we like cluster-admins and we dislike writing multiple pages of
+slightly different documentation.
+
+### Goals
+
+1. provide a way for cluster-admins to see all the customizeable routes in the cluster.
+2. provide a single API that is consistent for every route they need to configure
+3. provides a way to have an operator with scope limited permissions read the serving cert/key pairs
+4. allow a cluster-admin to specify a different DNS name.
+
+### Non-Goals
+
+1. provide a way to specify SNI serving certificates for operands.
+2. dictate how an operator must expose itself, the termination policy on a route, or how the secret is used.
+
+## Proposal
+
+```
+type IngressSpec struct {
+    // other fields
+
+    // CustomizeableRoutes is a list of routes that a cluster-admin wants to customize.  It is logically keyed by
+    // .spec.customizeableRoutes[index].{namespace,name}.
+    // To determine the set of possible keys, look at .status.customizeableRoutes where participating operators place
+    // current route status keyed the same way.
+    CustomizeableRoutes []CustomizeableRouteSpec
+}
+
+type IngressStatus struct {
+    // other fields
+
+    // CustomizeableRoutes is where participating operators place the current route status for routes which the cluster-admin
+    // can customize hostnames and serving certificates.
+    // How the operator uses that serving certificate is up to the individual operator.
+    CustomizeableRoutes []CustomizeableRouteStatus
+}
+
+type CustomizeableRouteSpec struct{
+    // namespace is the namespace of the route to customize.  It must be a real namespace.
+    Namespace string
+    // name is the *logical* name of the route to customize.  It does not have to be the actual name of a route resource.
+    // Keep in mind that this is your API for users to customize.  You could later rename the route, but you cannot rename
+    // this name.
+    Name string
+    // desiredHost is the name that a cluster-admin wants to specify
+    DesiredHost string
```

Comment:
This is in `spec`, so isn't "desired" redundant?

Reply:
what is a name? Be more precise. A name is two lines above already.

Reply:
We already have a similar logic in the `console-operator` where, in case of a custom route a hostname and a secret reference has to be specified.
Check: https://github.com/openshift/api/blob/master/operator/v1/types_console.go#L51-L63

Would drop the `desired` and just use hostname as we use in https://github.com/openshift/api/blob/master/operator/v1/types_console.go#L53